File: LeetCode/多线程/1114. 按序打印/cs/Foo.cs
using System.Threading;

public class Foo {
    private ManualResetEventSlim _mres2 = new ManualResetEventSlim();
    private ManualResetEventSlim _mres3 = new ManualResetEventSlim();

    public Foo()
    {    
    }

    public void First(Action printFirst) {
        
        // printFirst() outputs "first". Do not change or remove this line.
        printFirst();

        _mres2.Set();
    }

    public void Second(Action printSecond) {
        _mres2.Wait();
        _mres2.Reset();

        // printSecond() outputs "second". Do not change or remove this line.
        printSecond();

        _mres3.Set();
    }

    public void Third(Action printThird) {
        _mres3.Wait();
        _mres3.Reset();
        
        // printThird() outputs "third". Do not change or remove this line.
        printThird();
    }
}

File: LeetCode/多线程/1114. 按序打印/java/sync_method/Foo.java
class Foo {
    private boolean r1Done, r2Done;
    private Object lock;

    public Foo() {
        r1Done = false;
        r2Done = false;
        lock = new Object();
    }

    public void first(Runnable printFirst) throws InterruptedException {
        synchronized(lock) {
            // printFirst.run() outputs "first". Do not change or remove this line.
            printFirst.run();
            r1Done = true;
            lock.notifyAll();
        }
    }

    public void second(Runnable printSecond) throws InterruptedException {
        synchronized(lock) {
            while (!r1Done) {
                lock.wait();
            }
            // printSecond.run() outputs "second". Do not change or remove this line.
            printSecond.run();
            r2Done = true;
            lock.notifyAll();
        }
    }

    public void third(Runnable printThird) throws InterruptedException {
        synchronized(lock) {
            while (!r2Done) {
                lock.wait();
            }
            // printThird.run() outputs "third". Do not change or remove this line.
            printThird.run();
        }
    }
}